segmentation.py

```diff
@@ -0,0 +1,575 @@
+                not np.isnan(eta) and
+                np.isfinite(eta) and eta > n - 1.0):
+            msg = 'The shape parameter must be greater than one minus the'
+            msg += ' degrees of freedom.'
+            raise Exception(msg)
+
+        # Allocate space for storing the matrix of product statistics.
+        self.__prod = np.zeros([m + n, m + n])
+
+        # Initialize the statistics with the parameters of the prior
+        # distribution.
+        x = np.dot(omega, mu)
+        self.__prod[:m, :m] = omega
+        self.__prod[:m, m:] = x
+        self.__prod[m:, :m] = x.transpose()
+        self.__prod[m:, m:] = np.dot(np.transpose(mu), x) + eta * sigma
+        self.__weight = eta
+
+    def update(self, X, Y):
+        """Update the sufficient statistics given observed data.
+
+        The sufficient statistics represent the only parameters required to
+        describe the shape of the distribution. Due to conjugacy, the
+        sufficient statistics also form the hyper-parameters of the
+        distribution. In a Bayesian context, when the sufficient statistics are
+        referred to as 'hyper-parameters', it implies the sufficient statistics
+        reflect prior information about the distribution. After updating the
+        distribution with observed data, the sufficient statistics refer to the
+        posterior distribution.
+
+        """
+
+        # (Equation 5a, b)
+        #
+        #     | XX    XY |
+        #     | YX    YY |
+        #
+        if np.ndim(X) > 1:
+            k, m = np.shape(X)
+            x = np.dot(np.transpose(X), Y)
+
+            # Update the statistics given a block of data (in the following
+            # order: XX, XY, YX, YY)
+            self.__prod[:m, :m] += np.dot(np.transpose(X), X)
+            self.__prod[:m, m:] += x
+            self.__prod[m:, :m] += x.transpose()
+            self.__prod[m:, m:] += np.dot(np.transpose(Y), Y)
+            self.__weight += k
+
+        else:
+            m = np.size(X)
+            x = np.outer(X, Y)
+
+            # Update the statistics given a single datum.
+            self.__prod[:m, :m] += np.outer(X, X)
+            self.__prod[:m, m:] += x
+            self.__prod[m:, :m] += x.transpose()
+            self.__prod[m:, m:] += np.outer(Y, Y)
+            self.__weight += 1
+
+    def logconst(self):
+
+        m, n = self.__m, self.__n
+
+        # Note usage of the log-determinant 'trick':
+        #
+        #     log(det(A)) = 2*sum(log(diag(chol(A))))
+        #
+        d = np.diag(linalg.cholesky(self.__prod))
+        w = self.__weight
+
+        # Evaluate the log-normalization constant.
+        # (Equation 8)
+        return special.gammaln(0.5*(w - np.arange(n))).sum() - \
+               n * np.log(d[:m]).sum() - \
+               w * np.log(d[m:] / np.sqrt(w)).sum() - \
+               n * (0.5 * w) * np.log(0.5 * w)
+
+    def param(self):
+        """Return parameters of the posterior distribution."""
```

Comment:
Return the parameters of the posterior distribution.

All the information content of the data is summarized by the sufficient statistics. Hence the parameters of the posterior are a function of the sufficient statistics. This is a consequence of the conjugacy of the matrix-variate Gaussian-inverse-Gamma distribution and the linear model.